### ISSUE TITLE
`RFC3986` Path rule

### ISSUE THREAD

**Issue by st235** [declaresub/abnf#24]
It seems that `path` should also be reordered, right?

`path-abempty` always matches any string as it accepts **0 iterations**.

[This is a snippet](./src/abnf/grammars/rfc3986.py#L35) I am referring to:

```py
        "path = path-abempty / path-absolute / path-noscheme / path-rootless / path-empty",
        'path-abempty = *( "/" segment )',
        'path-absolute = "/" [ segment-nz *( "/" segment ) ]',
        'path-noscheme = segment-nz-nc *( "/" segment )',
        'path-rootless = segment-nz *( "/" segment )',
        "path-empty = 0pchar",
```

**Comment by st235** [comment]
`path-empty` also looks similar to `path-abempty`. Do you think it would be better to reorder the rules in the following order:

`path-absolute / path-noscheme / path-rootless / path-abempty` and remove `path-empty`?

**Comment by declaresub** [comment]
Did you encounter a problem because of this?   I think the current version of abnf has enough backtracking to handle this case.  Probably we need a test or two to ensure that this isn't a problem.

**Comment by st235** [comment]
> Did you encounter a problem because of this? I think the current version of abnf has enough backtracking to handle this case.

Hey, thank you for a swift reply! 
Actually, nope, don't have any problems so far, but I have not tested for this specific scenario.
It just looked peculiar when I read the `ABNF` notation, so I may be utterly wrong here.

**Comment by declaresub** [comment]
You're not wrong to make this observation; there are several rules in the various RFCs akin to this one.  I should add a test for this rule to ensure that the parser generated for RFC3886 does what one expects.

**Comment by st235** [comment]
I just see that you also reordered `dec-octet` and that is what I had in mind when I was googling for problems with `dec-octet`. This is actually how I found your repo in the first place 😅 

I am talking about this part of the code

```py
# alternatives reordered for correct matching.
'dec-octet = "25" %x30-35 / "2" %x30-34 DIGIT / "1" 2DIGIT / %x31-39 DIGIT / DIGIT',
```

Another problem within `RC3986` I am feeling is happening is `path` rule, and I was wondering if you had this issue with this rule or noticed something similar.

At least it seems that `path-abempty` and `path-empty`  are similar when matching for an **empty** string.

**Comment by st235** [comment]
@dekkers I actually figured it out, sorry for the confusion.
It seems that `path` rule itself is not used anywhere in the grammar.

Path children expressions are used in `heir-part` and `relative-part`

```py
...
'hier-part = "//" authority path-abempty / path-absolute / path-rootless / path-empty',
...
'relative-part = "//" authority path-abempty / path-absolute / path-noscheme / path-empty',
```

Though in these rules `path-abempty` is protected by terminal `//` symbols that help to select between alterations.

Really confusing definitions! So `URI`-rule should be safe and sound.


Though the issues hold for `path`-rule.

The snippet below (with empty string) always produces AST with one child: `path-abempty`, so it does not reach `path-empty`.

```py
src = ''
node, start = rfc3986.Rule('path').parse(src, 0)
print(str(node))
```

```bash
Node(name=path, children=[Node(name=path-abempty, children=[])])
```
